EGATE-84: events consolidator

jamalkm opened the pull request:
Added Activiti adapter which consumes events from RabbitMQ using Spring Cloud Stream and after transformation, sends the events to ActiveMQ using Spring JMS.

Commit:
Initial work

Initial work done for events consolidator

Commit:
Adapt project to Activiti's model

Commit:
Deleted old src folder

Commit:
Unmarshal list of events

jamalkm committed:
EGATE-84/EGATE-78: Added Activiti consolidator which consumes events from RabbitMQ using Spring Cloud Stream and after transformation, sends the events to ActiveMQ using Spring JMS.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Activiti/activiti-cloud-events-adapter-service?pullRequest=1) <br/>All committers have signed the CLA.

salaboy commented:
@cristina-sirbu can you please sign the CLA?

salaboy commented on activiti-cloud-events-adapter-service/src/main/java/org/activiti/cloud/events/adapter/jms/JmsProperties.java:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @author Jamal Kaabi-Mofrad
+ */
+@ConfigurationProperties(prefix = "messaging.to.activemq")
```
@jamalkm two things, we don't use @Author tags, please remove and second,
"messaging.to.activemq" sounds to broad, I would rather have something like: "activiti.cloud.events.adapter.queue.*"

jamalkm replied:
Sure, I'll remove the author tags (it's automatically added by IntelliJ using the Alfresco Java profile). 

@Salaboy We are using ActiveMQ topic rather than queue. So shouldn’t **activiti.cloud.events.adapter.queue.*** be **activiti.cloud.events.adapter.topic.*** ?

salaboy replied:
@jamalkm yes that sounds good to me.

rgauss reviewed (approved):
I know we discussed having the main logic contained in the `-adapter` project and the `-adapter-service` simply being a Spring Boot / Cloud wrapper around the `-adapter` dependency and hopefully we can evolve towards that with this PR just being the initial step.

I believe a Dockerfile and Helm chart will follow in a subsequent PR.

salaboy commented on activiti-cloud-events-adapter-service/src/main/java/org/activiti/cloud/events/adapter/streams/EventsAdapterConsumerChannel.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.messaging.SubscribableChannel;
+
+/**
+ * @author Jamal Kaabi-Mofrad
```
@jamalkm please remove author tags

salaboy commented on activiti-cloud-events-adapter-service/src/main/resources/application.properties:
```diff
@@ -0,0 +1,19 @@
+spring.cloud.stream.bindings.eventsAdapterConsumer.group=eventsAdapter
+spring.cloud.stream.bindings.eventsAdapterConsumer.contentType=application/json
+
+messaging.to.activemq.host=localhost
```
@jamalkm please refactor based on my previous comment about the name of the property.

salaboy reviewed (commented):
Small fixes, but it looks good. It will be good also to have an integration test with docker compose like the ones that we have in audit-service.

salaboy commented on activiti-cloud-events-adapter-service/pom.xml:
```diff
@@ -0,0 +1,156 @@
+        <dependency>
+            <groupId>org.apache.activemq.tooling</groupId>
+            <artifactId>activemq-junit</artifactId>
+            <version>5.15.5</version>
```
@jamalkm please let's add the version into a property so we can move that into dependency management later.

salaboy commented:
@rgauss @cristina-sirbu @jamalkm one thing that I don't see here is Spring Integrations (preferred) or Camel. If we are building an adapter layer, it will be way more useful if it can adapt and put the result in any other consume available. This looks very JMS centric, and I do understand that the current requirements are JMS, but with minimum effort we can add Spring Integrations and use the Spring Integrations built in features to write into a JMS topic. Does this make sense? 
Cheers

salaboy commented:
@rgauss for DockerFile and HELM charts.. let's use the other repository which will actually contain the service (this repo should only contain the starter). Also we are using Jenkins X for all other repositories, so following JX conventions is needed to keep uniformity across all Activiti repositories. But this is not needed right now as you mention.

rgauss commented:
> Spring Integrations (preferred) or Camel. If we are building an adapter layer, it will be way more useful if it can adapt and put the result in any other consume available.

As discussed in a Platform Services meeting, using Spring Integrations for the publishing side would be fine and @jamalkm and @cristina-sirbu will work on those changes.

rgauss commented:
> for DockerFile and HELM charts.. let's use the other repository which will actually contain the service (this repo should only contain the starter)

At some point we should converge on common naming conventions.  Currently the [guidelines](https://github.com/Alfresco/alfresco-anaxes-shipyard/blob/master/docs/adrs/0007-naming-and-versioning.md#github-repositories) recommend a suffix of `-deployment` on those type of repos, but we may end up revisiting that recommendation anyway (potentially with a single repo for code and deployment) so your suggested approach works for now.

jamalkm committed:
EGATE-84/EGATE-78: As per peer review:
                 - Removed author name
                 - Renamed activemq properties
                 - Used spring integration to send the JMS messages

jamalkm commented:
@Salaboy please see the latest commit for the requested changes.

Commit:
Added Spring Cloud Starter

salaboy commented on activiti-cloud-events-adapter-service/src/main/resources/META-INF/spring.factories:
```diff
@@ -0,0 +1,2 @@
+org.springframework.boot.autoconfigure.EnableAutoConfiguration=\
+  org.activiti.cloud.events.adapter.ActivitiCloudEventsAdapterService
```
@cristina-sirbu IF there is a configuration, you need to follow the conventions. It needs to be called ....AutoConfiguration

Commit:
Fix based on review comments

jx-activiti-cloud commented on activiti-cloud-starter-events-adapter/src/main/java/org/activiti/cloud/events/adapter/ActivitiCloudEventsAdapterAutoConfiguration.java:
```diff
@@ -23,10 +23,10 @@
 @EnableDiscoveryClient
 @EnableIntegration
-public class ActivitiCloudEventsAdapterService {
+public class ActivitiCloudEventsAdapterAutoConfiguration {
```
@cristina-sirbu this is not the auto-configuration.. a starter cannot have @SpringBootApplication. this is going to go to the other repo.. Please check how Audit is structured. We cannot have a main class here.

salaboy commented on activiti-cloud-starter-events-adapter/src/main/java/org/activiti/cloud/events/adapter/ActivitiCloudEventsAdapterAutoConfiguration.java:
```diff
@@ -23,10 +23,10 @@
 
     public static void main(String[] args) {
-        SpringApplication.run(ActivitiCloudEventsAdapterService.class, args);
+        SpringApplication.run(ActivitiCloudEventsAdapterAutoConfiguration.class, args);
```
@cristina-sirbu this is not the auto-configuration.. a starter cannot have @SpringBootApplication. this is going to go to the other repo.. Please check how Audit is structured. We cannot have a main class here. The AutoConfiguration is the one annotated with @Configuration

salaboy commented on activiti-cloud-events-adapter-service/src/test/java/org/activiti/cloud/events/adapter/ActivitiCloudEventsAdapterTest.java:
```diff
@@ -0,0 +1,121 @@
+        final String payload = getEventResourceAsString("aggregated-processStarted.json");
+        adapterConsumer.send(new GenericMessage<>(payload));
+        // wait for a few seconds to get all the messages
+        TimeUnit.SECONDS.sleep(5);
```
@cristina-sirbu we are using awaitability framework for waiting messages to be consumed.. can we use that here as well? https://github.com/Activiti/activiti-cloud-audit-service/blob/develop/activiti-cloud-starter-audit/src/test/java/org/activiti/cloud/starter/audit/tests/it/AuditServiceIT.java#L95

Notice also that we have our IT tests inside the starter project to simulate a real application using the starter as an add-on, hence I would move this class to the starter

salaboy commented on activiti-cloud-events-adapter-service/src/test/resources/application.properties:
```diff
@@ -0,0 +1,18 @@
+logging.level.org.activiti.cloud.events.adapter=debug
```
@cristina-sirbu same this file, which seems to be used to test the streams with real queues, this might be better in the starter project.

salaboy commented on activiti-cloud-starter-events-adapter/src/main/java/org/activiti/cloud/events/adapter/ActivitiCloudEventsAdapterAutoConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+
+@SpringBootApplication
+@EnableDiscoveryClient
+@EnableIntegration
```
@cristina-sirbu as I mentioned before this class is not the AutoConfiguration and we cannot have a @SpringBootApplication inside the starter. We can only have one inside the tests for test purposes.

Commit:
Addressing review comments

salaboy commented:
@jamalkm @cristina-sirbu great work! well done! I will merge this now so we can move to the docker image in the activiti-cloud-events-adapter repository
